File: code/2025/04/10/1143. 最长公共子序列.cpp
class Solution {
public:
	int longestCommonSubsequence(string s, string t) {
		int n = s.size(), m = t.size();
		n++, m++;
		std::vector<int> f(m);
		for (int i = 1; i < n; i++) {
			std::vector<int> nf(m);
			for (int j = 1; j < m; j++) {
				if (s[i - 1] == t[j - 1]) {
					nf[j] = f[j - 1] + 1;
				} else {
					nf[j] = std::max(nf[j - 1], f[j]);
				}
			}
			f = nf;
		}
		return f.back();
	}
};

File: code/2025/04/10/1143. 最长公共子序列.py
class Solution:
    def longestCommonSubsequence(self, s: str, t: str) -> int:
        n, m = len(s), len(t)
        @cache
        def dfs(i, j):
            if i < 0 or j < 0:
                return 0
            if s[i] == t[j]:
                return dfs(i - 1, j - 1) + 1
            return max(dfs(i - 1, j), dfs(i, j - 1))
        return dfs(n - 1, m - 1)